File: Design-Patterns-Using-CSharp/Singleton/DatabaseSingleton.cs
using System;
using System.Data;
using System.Data.SQLite;

public class DatabaseSingleton
{
    private static DatabaseSingleton instance;
    private IDbConnection connection;
    private IDbCommand command;

    private DatabaseSingleton()
    {
        try
        {
            Console.WriteLine("Database created");
            connection = new SQLiteConnection("Data Source=db.sqlite3;");
            connection.Open();
            command = connection.CreateCommand();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.ToString());
        }
    }

    public static DatabaseSingleton GetInstance()
    {
        if (instance == null)
        {
            instance = new DatabaseSingleton();
        }
        return instance;
    }

    public void CreateTable()
    {
        try
        {
            command.CommandText = "CREATE TABLE IF NOT EXISTS students (id INTEGER, name TEXT);";
            command.ExecuteNonQuery();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.ToString());
        }
    }

    public void AddData(int id, string name)
    {
        try
        {
            command.CommandText = $"INSERT INTO students (id, name) VALUES ({id}, '{name}');";
            command.ExecuteNonQuery();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.ToString());
        }
    }

    public void Display()
    {
        try
        {
            command.CommandText = "SELECT * FROM students;";
            using (IDataReader reader = command.ExecuteReader())
            {
                while (reader.Read())
                {
                    int studentId = Convert.ToInt32(reader["id"]);
                    string studentName = Convert.ToString(reader["name"]);
                    Console.WriteLine($"ID: {studentId}, Name: {studentName}");
                }
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e.ToString());
        }
    }

    // Client code
    public static void Main(string[] args)
    {
        DatabaseSingleton db1 = DatabaseSingleton.GetInstance();
        DatabaseSingleton db2 = DatabaseSingleton.GetInstance();

        Console.WriteLine($"Database Objects DB1: {db1}");
        Console.WriteLine($"Database Objects DB2: {db2}");

        db1.CreateTable();
        db1.AddData(1, "john");
        db2.AddData(2, "smith");

        db1.Display();
    }
}

File: Design-Patterns-Using-Ruby/Singleton/DatabaseSingleton.rb
require 'sqlite3'

class DatabaseSingleton
  @@instance = nil

  def self.instance
    @@instance ||= new
  end

  def initialize
    begin
      puts "Database created"
      @connection = SQLite3::Database.new('db.sqlite3')
    rescue SQLite3::Exception => e
      puts "Exception occurred: #{e}"
    end
  end

  def create_table
    begin
      @connection.execute("CREATE TABLE IF NOT EXISTS students (id INTEGER, name TEXT);")
    rescue SQLite3::Exception => e
      puts "Exception occurred: #{e}"
    end
  end

  def add_data(id, name)
    begin
      query = "INSERT INTO students (id, name) VALUES (?, ?);"
      @connection.execute(query, id, name)
    rescue SQLite3::Exception => e
      puts "Exception occurred: #{e}"
    end
  end

  def display
    begin
      result_set = @connection.execute("SELECT * FROM students;")
      result_set.each do |row|
        puts "ID: #{row[0]}, Name: #{row[1]}"
      end
    rescue SQLite3::Exception => e
      puts "Exception occurred: #{e}"
    end
  end
end

# Client code
db1 = DatabaseSingleton.instance
db2 = DatabaseSingleton.instance

puts "Database Objects DB1: #{db1}"
puts "Database Objects DB2: #{db2}"

db1.create_table
db1.add_data(1, 'john')
db2.add_data(2, 'smith')

db1.display

=begin 
Database created
Database Objects DB1: #<DatabaseSingleton:0x00007ff1959146f8>
Database Objects DB2: #<DatabaseSingleton:0x00007ff1959146f8>
ID: 1, Name: john
ID: 2, Name: smith
=end